redisinsight/api/src/modules/bulk-actions/bulk-import.service.ts

```diff
@@ -0,0 +1,119 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { Readable } from 'stream';
+import * as readline from 'readline';
+import { wrapHttpError } from 'src/common/utils';
+import { UploadImportFileDto } from 'src/modules/bulk-actions/dto/upload-import-file.dto';
+import { DatabaseConnectionService } from 'src/modules/database/database-connection.service';
+import { ClientMetadata } from 'src/common/models';
+import { splitCliCommandLine } from 'src/utils/cli-helper';
+import { BulkActionSummary } from 'src/modules/bulk-actions/models/bulk-action-summary';
+import { IBulkActionOverview } from 'src/modules/bulk-actions/interfaces/bulk-action-overview.interface';
+import { BulkActionStatus, BulkActionType } from 'src/modules/bulk-actions/constants';
+
+const BATCH_LIMIT = 10_000;
+
+@Injectable()
+export class BulkImportService {
+  private logger = new Logger('BulkImportService');
+
+  constructor(
+    private readonly databaseConnectionService: DatabaseConnectionService,
+  ) {}
+
+  private async executeBatch(client, batch: any[]): Promise<BulkActionSummary> {
+    const result = new BulkActionSummary();
+    result.addProcessed(batch.length);
+
+    try {
+      if (client?.isCluster) {
+        await Promise.all(batch.map(async ([command, args]) => {
+          try {
+            await client.call(command, args);
+            result.addSuccess(1);
+          } catch (e) {
+            result.addFailed(1);
+          }
+        }));
+      } else {
+        (await client.pipeline(batch).exec()).forEach(([err]) => {
+          if (err) {
+            result.addFailed(1);
+          } else {
+            result.addSuccess(1);
+          }
+        });
+      }
+    } catch (e) {
+      this.logger.error('Unable to execute batch of commands', e);
+      result.addFailed(batch.length);
+    }
+
+    return result;
+  }
+
+  /**
+   * @param clientMetadata
+   * @param dto
+   */
+  public async import(clientMetadata: ClientMetadata, dto: UploadImportFileDto): Promise<IBulkActionOverview> {
+    const result: IBulkActionOverview = {
+      id: 'empty',
+      databaseId: clientMetadata.databaseId,
+      type: BulkActionType.Import,
+      summary: {
+        processed: 0,
+        succeed: 0,
+        failed: 0,
+        errors: [],
+      },
+      progress: null,
+      filter: null,
+      status: BulkActionStatus.Completed,
+      duration: Date.now(),
```

Comment:
Maybe will be better to create another constant startBulkActionTime or something else, it will be more clear i guess. And in the end result.duration = Date.now() - startBulkActionTime. But it is up to you.

Reply:
sure. if this will be more easy to understand